raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+        bytes reward_proof_signature;
+
+        // Nonce of the most recently provided BP
+        uint256 nonce;
+
+        // Address of the Raiden Node that was monitored
+        // This is also the address that has the reward deducted from its deposit
+        address reward_sender_address;
+    }
+
+    /*
+     *  Events
+     */
+
+    //event MonitoringServiceNewDeposit(address sender, address receiver, uint amount);
+    event NewDeposit(address sender, address receiver, uint amount);
+    event RegisteredMonitoringService(address monitoring_service);
+    event NewBalanceProofReceived(
+        bytes reward_proof_signature,
+        uint256 nonce,
+        address ms_address,
+        address raiden_node_address
+    );
+    event RewardClaimed(address ms_address, uint amount, bytes32 channel_id);
+    event MonitoringServiceRegistered(address ms_address);
+    event MonitoringServiceDeregistered(address ms_address);
+    event Withdrawn(address account, uint amount);
+
+    /*
+     *  Modifiers 
+     */
+
+    modifier isMonitor(address _ms_address) {
+        require(registered_monitoring_services[_ms_address]);
+        _;
+    }
+
+    modifier isNotMonitor(address raiden_node_address) {
+        require(!registered_monitoring_services[raiden_node_address]);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    /// @param _minimum_deposit The minimum value that a MS must deposit to be registered
+    constructor(
+        address _token_address,
+        uint256 _minimum_deposit
+    )
+        public
+    {
+        require(_token_address != 0x0);
+        require(_minimum_deposit > 0);
+        require(contractExists(_token_address));
+
+        minimum_deposit = _minimum_deposit;
+
+        token = Token(_token_address);
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+    }
+
+    /// @notice Deposit tokens used to either pay MSs or to regsiter as MS
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefitting from the deposit
+    /// @param amount The amount of tokens to be depositted
+    function deposit(address beneficiary, uint amount) public {
+        require(amount > 0);
+
+        // This also allows for registered MSs to deposit and use other MSs
+        balances[beneficiary] += amount;
+
+        // Transfer the deposit to the smart contract
+        require(token.transferFrom(msg.sender, address(this), amount));
+
+        emit NewDeposit(msg.sender, beneficiary, amount);
+    }
+
+    /// @notice Deposit and register as Monitoring Service in one function call
+    /// Can be called once and is only callable by addresses not already registered as MS
+    function depositAndRegisterMonitoringService() isNotMonitor(msg.sender) public {
```

Comment:
We should not need this. Depositing as an MS should check the minimum deposit and having a balance in the contract should be enough to be considered registered.